encoding/codecv4/codecv4.go

```diff
@@ -0,0 +1,534 @@
+package codecv4
+
+/*
+#include <stdint.h>
+char* compress_scroll_batch_bytes(uint8_t* src, uint64_t src_size, uint8_t* output_buf, uint64_t *output_buf_size);
+*/
+import "C"
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+	"unsafe"
+
+	"github.com/scroll-tech/go-ethereum/accounts/abi"
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/crypto"
+	"github.com/scroll-tech/go-ethereum/crypto/kzg4844"
+	"github.com/scroll-tech/go-ethereum/log"
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/da-codec/encoding/codecv1"
+	"github.com/scroll-tech/da-codec/encoding/codecv3"
+)
+
+// MaxNumChunks is the maximum number of chunks that a batch can contain.
+const MaxNumChunks = codecv3.MaxNumChunks
+
+// DABlock represents a Data Availability Block.
+type DABlock = codecv3.DABlock
+
+// DAChunk groups consecutive DABlocks with their transactions.
+type DAChunk = codecv3.DAChunk
+
+// DABatch contains metadata about a batch of DAChunks.
+type DABatch struct {
+	// header
+	Version              uint8          `json:"version"`
+	BatchIndex           uint64         `json:"batch_index"`
+	L1MessagePopped      uint64         `json:"l1_message_popped"`
+	TotalL1MessagePopped uint64         `json:"total_l1_message_popped"`
+	DataHash             common.Hash    `json:"data_hash"`
+	BlobVersionedHash    common.Hash    `json:"blob_versioned_hash"`
+	ParentBatchHash      common.Hash    `json:"parent_batch_hash"`
+	LastBlockTimestamp   uint64         `json:"last_block_timestamp"`
+	BlobDataProof        [2]common.Hash `json:"blob_data_proof"`
+
+	// blob payload
+	blob *kzg4844.Blob
+	z    *kzg4844.Point
+}
+
+// NewDABlock creates a new DABlock from the given encoding.Block and the total number of L1 messages popped before.
+func NewDABlock(block *encoding.Block, totalL1MessagePoppedBefore uint64) (*DABlock, error) {
+	return codecv3.NewDABlock(block, totalL1MessagePoppedBefore)
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	return codecv3.NewDAChunk(chunk, totalL1MessagePoppedBefore)
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch, enableEncode bool) (*DABatch, error) {
```

Comment:
looks like `enableEncode` this variable actual means whether to enable zstd compression, right? Better to change the name to `enableCompress`.

Reply:
makes sense. changed in [c6af3bb](https://github.com/scroll-tech/da-codec/pull/23/commits/c6af3bbe7068da2b356509f2cb9eaf6c8d514bdf).